### ISSUE TITLE
SimpleForm isDirty is always true if ArrayInput is used

### ISSUE THREAD

**Issue by AlexDroll** [marmelab/react-admin#9661]
**What you were expecting:**
* if no input is changed the form should still be not dirty

**What happened instead:**
* `isDirty` property using the `useFormState()` hook from react-hook-form is set to `true`
* for the initial rendering `isDirty` is still `false` but becomes `true` in the following renderings (see console output in the sandbox)
* the `dirtyFields` are still empty as expected
* in the `defaultValues` the source of the ArrayInput `groups` disappears after the first rendering (see console output in the sandbox)

**Steps to reproduce:**
* open the code sandbox
* open the console 
* reload the page

**Related code:**
* Code sandbox with minimal example: [here](https://codesandbox.io/p/devbox/simple-form-isdirty-is-always-true-for-an-array-input-5fgvy7)

**Other information:**
* this issue is also related with react-hook-form updates
* the `isDirty` worked with react-hook-form@7.45.4 
* The issue with the disappearing default value existed already for react-hook-form@7.45.4

**Environment**

* React-admin version: 4.16.10
* Last version that did not exhibit the issue (if applicable):
* React version: 18.2.0
* Browser: firefox, chrome

**Comment by slax57** [comment]
Reproduced, thanks!
Appreciate the great sandbox btw :blush: 

Until this is fixed, you can use the same workaround we used in `SaveButton.tsx`:

```ts
    const { dirtyFields, isValidating, isSubmitting } = useFormState();
    // useFormState().isDirty might differ from useFormState().dirtyFields (https://github.com/react-hook-form/react-hook-form/issues/4740)
    const isDirty = Object.keys(dirtyFields).length > 0;
```

**Comment by AlexDroll** [comment]
Yes. We use the same workaround. 🙏 

Just one hint, this only works partially. If I change one value in the array input and then revert the changes the field is still listed in the `dirtyFields`. As a consequence also the `SaveButton` is enabled.

**Comment by Nilegfx** [comment]
I am not sure if I should open separate issue or just comment in this issue. so I will start with a comment and maybe separate it if needed. 

I am facing relatively same inconveniences with ArrayInput and dirty behaviour. in my case, I am using zod resolver.

**What you were expecting:** 
1. the form is not dirty
2. clicking the plus button adds an element and makes the form dirty
3. clicking the save button (without properly filling the inputs) the form should report the errors and the save button should stay enabled (as the array input still has an element added and the logic of save button rely on dirty state not validation state)
4. properly fill the input and click save should successfully submit the form

**What happened instead:**

https://user-images.githubusercontent.com/1198975/a06ba812-72af-4f2b-8624-fadd249ff55d

**Steps to reproduce:**

here are the two stackblitz that reproduces the two behaviours
on the right side (browser) navigate to /#/posts/create

* [old-working (with react-hook-form 7.50.1)](https://stackblitz.com/edit/vitest-dev-vitest-m9paul?file=package.json)
* [broken (with latest react-hook-form)](https://stackblitz.com/edit/vitest-dev-vitest-uxogjl?file=package.json)
there is no difference between the two repo other than the resolution key in the package.json


**Other information:**
_note that I fixed the react-hook-form version to `7.50.1` as the next release [`7.51.0`](https://github.com/react-hook-form/react-hook-form/releases/tag/v7.51.0) is the one that introduced my issue_

**Comment by Nilegfx** [comment]
I think [this issue](https://github.com/react-hook-form/react-hook-form/issues/11719) on react-hook-form side is related
